integration_test/cases/repo.exs

```diff
@@ -953,6 +953,73 @@ defmodule Ecto.Integration.RepoTest do
     assert custom.bid == bid2
   end
 
+  @tag :ph
+  test "Repo.insert_all PH main" do
+    now = Date.utc_today
+    posts = [
+      %{title: {:placeholder, :foo}, visits: 111, posted: {:placeholder, :today}},
+      %{title: {:placeholder, :foo}, visits: 222, posted: {:placeholder, :today}},
+    ]
+    TestRepo.insert_all(Post, posts, placeholders: %{foo: "Title", today: now})
+
+    query = from(p in Post, select: {p.title, p.visits, p.posted})
+    assert [{"Title", 111, now}, {"Title", 222, now}] == TestRepo.all(query)
+  end
+
+  @tag :ph
+  test "Repo.insert_all fills in placeholders" do
+    TestRepo.insert_all(Post, [%{title: {:placeholder, :foo}}], placeholders: %{foo: "Title"})
+
+    query = from(p in Post, select: p.title)
+    assert ["Title"] == TestRepo.all(query)
+  end
+
+  @tag :ph
+  test "Repo.insert_all accepts non atom placeholder keys" do
+    placeholders = %{10 => "integer key", {:foo, :bar} => "tuple key"}
+    entries = [%{title: {:placeholder, 10}}, %{title: {:placeholder, {:foo, :bar}}}]
+    TestRepo.insert_all(Post, entries, placeholders: placeholders)
+
+    query = from(p in Post, select: p.title)
+    assert ["integer key", "tuple key"] == TestRepo.all(query)
+  end
+
+  @tag :ph
+  test "Repo.insert_all fills in placeholders with keyword list entries" do
+    TestRepo.insert_all(Post, [[title: {:placeholder, :foo}]], placeholders: %{foo: "Title"})
+
+    query = from(p in Post, select: p.title)
+    assert ["Title"] == TestRepo.all(query)
+  end
+
+  @tag :ph
+  test "Repo.insert_all throws when placeholder key is not found" do
+    assert_raise KeyError, fn ->
+      TestRepo.insert_all(Post, [%{title: {:placeholder, :bad_key}}], placeholders: %{foo: "Title"})
+    end
+  end
+
+  @tag :ph
+  test "Repo.insert_all throws when placeholder key is used for different types" do
+    placeholders = %{uuid_key: Ecto.UUID.generate}
+    ph_key = {:placeholder, :uuid_key}
+    entries = [%{bid: ph_key, title: ph_key}]
+
+    assert_raise ArgumentError, fn ->
+      TestRepo.insert_all(Post, entries, placeholders: placeholders)
+    end
+  end
+
+  @tag :ph
+  test "Repo.insert_all throws when placeholder key is used with invalid types" do
+    placeholders = %{string_key: "foo"}
+    entries = [%{visits: {:placeholder, :string_key}}]
+
+    assert_raise Ecto.ChangeError, fn ->
+      TestRepo.insert_all(Post, entries, placeholders: placeholders)
+    end
+  end
```

Comment:
Tests that check it raises should not be integration tests. Instead, please move them to test/ecto/repo_test.exs (our unit tests).